File: models/api_key.go
package models

import (
	"strings"
	"time"
)

// Permisos soportados por una API key. "admin" implica read+write y además
// permite operaciones de administración de buckets ajenos.
const (
	PermissionRead  = "read"
	PermissionWrite = "write"
	PermissionAdmin = "admin"
)

// APIKey representa una credencial emitida para un usuario. El secreto en
// texto plano nunca se persiste: sólo se guarda su hash bcrypt (KeyHash).
type APIKey struct {
	ID          string     `json:"id" db:"id"`
	UserID      string     `json:"user_id" db:"user_id"`
	KeyHash     string     `json:"-" db:"key_hash"`
	Permissions string     `json:"permissions" db:"permissions"`
	CreatedAt   time.Time  `json:"created_at" db:"created_at"`
	LastUsed    *time.Time `json:"last_used,omitempty" db:"last_used"`
}

// HasPermission indica si la key tiene el permiso solicitado (admin implica
// todos los demás).
func (k *APIKey) HasPermission(perm string) bool {
	for _, p := range strings.Split(k.Permissions, ",") {
		p = strings.TrimSpace(p)
		if p == PermissionAdmin || p == perm {
			return true
		}
	}
	return false
}

// CreateAPIKeyRequest es el payload de POST /api/auth/keys.
type CreateAPIKeyRequest struct {
	UserID      string `json:"user_id"`
	Permissions string `json:"permissions"` // ej: "read,write"
}

// CreateAPIKeyResponse devuelve la key en texto plano UNA sola vez.
type CreateAPIKeyResponse struct {
	ID          string `json:"id"`
	UserID      string `json:"user_id"`
	Key         string `json:"key"`
	Permissions string `json:"permissions"`
	CreatedAt   string `json:"created_at"`
}

File: core/Entities/ApiKey.cs
namespace GabichoStorage.Api.Entities;

public static class Permission
{
    public const string Read = "read";
    public const string Write = "write";
    public const string Admin = "admin";

    public static readonly HashSet<string> Valid = [Read, Write, Admin];
}

/// <summary>
/// Credencial emitida para un usuario. El secreto en texto plano nunca se
/// persiste: sólo se guarda su hash bcrypt (KeyHash).
/// </summary>
public class ApiKey
{
    public Guid Id { get; set; }

    public required string UserId { get; set; }

    public required string KeyHash { get; set; }

    /// <summary>Lista separada por coma, ej. "read,write". "admin" implica todos los demás.</summary>
    public required string Permissions { get; set; }

    public DateTimeOffset CreatedAt { get; set; }

    public DateTimeOffset? LastUsed { get; set; }

    /// <summary>
    /// Indica si esta key tiene el permiso solicitado (admin implica todos
    /// los demás).
    /// </summary>
    public bool HasPermission(string permission)
    {
        foreach (var p in Permissions.Split(',', StringSplitOptions.TrimEntries))
        {
            if (p == Permission.Admin || p == permission)
            {
                return true;
            }
        }
        return false;
    }
}
